drivers/stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,291 @@
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	*is_moving = k_work_delayable_is_pending(&data->stepper_dwork);
+	LOG_DBG("Is Motor Moving %d", *is_moving);
+	return 0;
+}
+
+static int gpio_stepper_set_max_velocity(const struct device *dev, uint32_t velocity)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	if (velocity == 0) {
+		LOG_ERR("Velocity cannot be zero");
+		return -EINVAL;
+	}
+
+	K_SPINLOCK(&data->lock) {
+		data->delay_in_us = USEC_PER_SEC / velocity;
+	}
+	LOG_DBG("Setting Motor Speed to %d", velocity);
+	return 0;
+}
+
+static int gpio_stepper_enable_constant_velocity_mode(const struct device *dev,
+						      const enum stepper_direction direction,
+						      const uint32_t value)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		data->run_mode = STEPPER_VELOCITY_MODE;
+		data->direction = direction;
+		k_work_reschedule(&data->stepper_dwork, K_NO_WAIT);
+	}
+	return 0;
+}
+
+static int gpio_stepper_get_micro_step_res(const struct device *dev,
+					   enum micro_step_resolution *micro_step_res)
+{
+	ARG_UNUSED(dev);
+	*micro_step_res = STEPPER_FULL_STEP;
+	return 0;
+}
+
+static int gpio_stepper_motor_controller_init(const struct device *dev)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+	const struct gpio_stepper_motor_controller_config *config = dev->config;
+
+	data->dev = dev;
+	LOG_INF("Initializing %s gpio_stepper_motor_controller with %d pin", dev->name,
+		config->number_of_control_pins);
+	for (uint8_t n_pin = 0; n_pin < config->number_of_control_pins; n_pin++) {
+		gpio_pin_configure_dt(&config->control_pins[n_pin], GPIO_OUTPUT_INACTIVE);
+	}
+	k_work_init_delayable(&data->stepper_dwork, stepper_work_step_handler);
+	return 0;
+}
+
+#define GPIO_STEPPER_DEVICE_DATA_DEFINE(child)                                                     \
+	static struct gpio_stepper_motor_controller_data gpio_stepper_motor_controller_data_##child;
+
+#define GPIO_STEPPER_DEVICE_CONFIG_DEFINE(child)                                                   \
+	static const struct gpio_dt_spec gpio_stepper_motor_control_pins_##child[] = {             \
+		DT_FOREACH_PROP_ELEM_SEP(child, gpios, GPIO_DT_SPEC_GET_BY_IDX, (,)),              \
+	};                                                                                         \
+	static const struct gpio_stepper_motor_controller_config                                   \
+		gpio_stepper_motor_controller_config_##child = {                                   \
+			.control_pins = gpio_stepper_motor_control_pins_##child,                   \
+			.number_of_control_pins =                                                  \
+				ARRAY_SIZE(gpio_stepper_motor_control_pins_##child),               \
+			.gpio_stepper_lookup_table =                                               \
+				COND_CODE_1(DT_PROP(child, gpios_IDX_3_EXISTS),                    \
```

Comment:
What is property `gpio_IDX_3_EXIT`? Can't found it in bindings. Also I don't believe that this code is working. All the Zephyr DT_*/DT_INST_* macros expect values in lower case, never upper case or in this case mixed. Maybe I don't know all.

At least, expand the binding with an required (or not?) property and explain the technical meaning for users / developers.

BTW, what reason I currently can't see requires an build-time change on a complete const and static LUT? For my point of view the driver needs only one `static const` qualified table of some quadruple integer values. See: https://github.com/tiacsys/bridle/blob/eval/zephyr-prs/stepper-motor/drivers/stepper/stepper_gpio.c#L55:

- full-step: https://github.com/tiacsys/bridle/blob/eval/zephyr-prs/stepper-motor/drivers/stepper/stepper_gpio.c#L64
- half-step: https://github.com/tiacsys/bridle/blob/eval/zephyr-prs/stepper-motor/drivers/stepper/stepper_gpio.c#L85

Reply:
> What is property `gpio_IDX_3_EXISTS`

So if you have 4 wire connection, then this property will exist in the device tree allowing me to select a certain gpio lookup table. In the future this can be expanded to two or five wire connection. 

> Also I don't believe that this code is working

Well this code is actually working, you can test it on any of the boards, i am working with nucleo_g071rb.

> For my point of view the driver needs only one `static const` qualified table of some quadruple integer values.

This lookup table will change depending on the number of wires. You can look it up in the `Arduino Stepper Library`.

Reply:
> nonsense, you never need a separate API per child motor - read, understand and (with pleasure) derive from our code: https://github.com/tiacsys/bridle/blob/eval/zephyr-prs/stepper-motor/drivers/stepper/stepper_gpio.c
> 
> Yes, you come hopefully over to the conclusion, that the current API has many deficits, e.g. can't handle multiple motors per driver instance (see my hints today to the TMC5041). Also here, read, understand and (with pleasure) derive from our code: https://github.com/tiacsys/bridle/blob/eval/zephyr-prs/stepper-motor/include/zephyr/drivers/stepper.h:

I'm sure you only want to help, but the way you are phrasing sounds somewhat arrogant, I'd suggest rephrasing those comments, not only these particular, since this is not the first ones, as said previously by @DaAwesomeP, and surely noticed by every other reviewer.

Reply:
> > What is property `gpio_IDX_3_EXISTS`
> 
> So if you have 4 wire connection, then this property will exist in the device tree allowing me to select a certain gpio lookup table. In the future this can be expanded to two or five wire connection.

Okay, but what is with the syntax. I've never seen such construct in Zephyr code. What does the code do and where is it recommended to do so?

> 
> > Also I don't believe that this code is working
> 
> Well this code is actually working, you can test it on any of the boards, i am working with nucleo_g071rb.
> 
> > For my point of view the driver needs only one `static const` qualified table of some quadruple integer values.
> 
> This lookup table will change depending on the number of wires. You can look it up in the `Arduino Stepper Library`.

Hm, would be a nice feature. Only I can't see any advantages in this practice. The only variable from perspective of hardware (device tree) are the 4 GPIO pin assignments. In correct order (that has to be documented in the binding, [see](https://github.com/tiacsys/bridle/blob/eval/zephyr-prs/stepper-motor/dts/bindings/stepper/gpio-steppers.yaml#L61)) the in-driver LUT matches every time.

When sometime in the future a 5-wire (uni-polar ?) system must be supported, you'll be able to extend the static and const in-driver LUT and for better performance use *DT_/DT_INST_* macros to pick-up only relevant LUT elements depending what device tree is telling the build system. So, it sounds for me a little bit over engineered to create dependencies for the LUT to device tree notations. That's unusually and I think also difficult to document.

Reply:
> @rexut I'm sure you only want to help, but the way you are phrasing sounds somewhat arrogant, I'd suggest rephrasing those comments, not only these particular, since this is not the first ones, as said previously by @DaAwesomeP, and surely noticed by every other reviewer.

Sorry, but technically and conceptually wrong code is not going to help the Zephyr project. If Zephyr is to be used in the safety critical market, a review has to hurt sometimes. Most of us are not natives and try to communicate our thoughts to each other in a foreign language. If the German term “Nonsens” for absolutely unusual code sequences and not explicitly pointed out as a possibility by the Zephyr documentation is considered a personal attack in other parts of the world, then I apologize. Please give me the correct vocabulary for such code. I have to name it somehow, otherwise the PR cannot be turned to positive.

Reply:
I would just keep those commentaries out. Suggest your changes and explain why they are more advantageous than the other. If it seems to be too pushy, then maybe it is. We totally understand your points and the technical needs you addressed, hence this is going to be discussed in the next Arch WG. It is just the phrasing that sounded too harsh more than a few times. Rest assured you got valid points to be discussed!